docs/rfcs/PipelineComposition.md

```diff
@@ -0,0 +1,170 @@
+# RFC: High Level MLIR Pass Pipeline Composition
+
+Ivan Butygin, Renato Golin
+
+## Summary
+
+We propose a new API to build a dependencies-based graph of pass pipelines.
+
+## Motivation
+
+At this point MLIR pass infrastruture only allows to define a linear sequence of passes. While this approach works
+for simple cases, more complex compiler pipelines may require more control over passes execution order.
+
+Two main usecases we are considering:
+* Extensibility. We want to have the abilty for users to extend existing pipeline by inserting theirs custom passes
+into various places while reusing most of the exiting pipeline. With current approach the most common way way to achieve this is
+for pipeline developer to add a fixed 'extension point' during initial pipeline design.
+* Dynamic pipeline control flow. It's often required to have an ability select specific sequence of passes based on some info which
+is only becomes available durin compilation process. It can be either branching to some separate sequnces of passes based on selected codegen path
+or running sequence of passes in the loop until some runtime condition is reached.
+
+## Proposal
+
+This API also allows a dynamic control flow between pipelines in graph, controlled by passes inside said pipelines.
+
+## New APIs
+
+### PipelineGraph
+```C++
+class PipelineGraph {
+public:
+    void registerPipeline(
+        StringRef name,
+        ArrayRef<StringRef> predecessors,
+        ArrayRef<StringRef> successors,
+        ArrayRef<StringRef> jumpTargets,
+        std::function<void(OpPassManager &)> populateFunc);
+
+    FailureOr<PipelineSchedule> createPipelineSchedule(raw_ostream &errorStream) const;
+};
+```
+`PipelineGraph` is main entry point for this new API.
+User is adding pipelines into graph via `registerPipeline` function, each pipeline have the following properties:
+* `name` - Name of thepipeline
+* `predecessors` - List of names of pipelines which must be run before current pipeline.
+* `successors` - List of the names of pipelines which must be run after current pipeline.
+* `jumpTargets` - List of the names of pipelines to which we can dynamically jump after current pipeline.
+* `populateFunc` - Callback to populate pipeline with passes.
+
+After user populated the graph object they must call `createPipelineSchedule` method to compile the resulted graph into runnable schedule.
+`createPipelineSchedule` will build a DAG from pipelines dependencies provided by user, and will try to get linear execution order to satify these dependencies.
+
+If two pipelines doesn't have direct and indirect dependencies, order in which they will be executed is not specified, but stable.
```

Comment:
I think that's asking too much of this framework. I'd say "stability is depending on the passes accepting canonical forms from each other", and we make sure we always run canonicalization between DAG nodes.

Reply:
Here I only meant order of pipelines/passes is stable regardless of in which order `registerPipeline`s were called (in POC impl I'm just sorting by pipeline name first to make it stable), but yes, I can remove this for more implementation freedom.